### ISSUE TITLE
Workspace should clear open documents that are removed during OnProjectReloaded

### ISSUE THREAD

**Issue by KirillOsenkov** [dotnet/roslyn#35963]
When `Workspace.OnProjectReloaded()` is called with a smaller set of documents (effectively removing documents from a project), and if any of those documents were open, the workspace doesn't properly stop tracking those open documents unless the host explicitly calls `OnDocumentRemoved()` beforehand.

The logic in `Workspace.OnProjectRemoved` calls `ClearProjectData()`, which takes care of this. However OnProjectReloaded doesn't have this. It only has the logic to transplant open documents to the new snapshot.

I think an else block is missing here:
https://github.com/dotnet/roslyn/blob/676d4721390ed7419eea3c0f3fdb4a8d6725e7cc/src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs#L736

And inside the else block it should probably call `ClearOpenDocument(docId)`

**Comment by KirillOsenkov** [comment]
@mavasani @jasonmalinowski basically I think if a document was present in the old project and open, and the new project doesn't have the document, we should close the document, as if the host had called `OnDocumentRemoved`

**Comment by CyrusNajmabadi** [comment]
I'm usually against this sort of thing.  It often feels (to me at least) like this is the Workspace trying to paper over callers calling in in a manner that papers over bugs on their side.  i.e. it's better to be stricter and to impose and catch more restrictive requirements (i.e. an `open` must be paired with a `close`) versus trying to catch this sort of thing and be resilient to it.

What do you think @KirillOsenkov ?  I know that can be harder on the host.  But i think it helps ensure that all parties are tracking data properly and also expecting clear and consistent relationships between operations.

**Comment by jasonmalinowski** [comment]
This just sounds like a bug per @KirillOsenkov's description. Prior to 16.0 it was a requirement from the workspace that we'd have to close all the documents from a project before removing it; I removed that entirely since:

1. It wasn't actually resulting in any way to discover bugs like @CyrusNajmabadi is concerned about. The "fix" everybody did was just loop through their open files and call Close prior to calling Remove. The workspace just pushed the papering over to the host.
2. It's faster, since to remove a project you're not running through a bunch of WorkspaceChanged events to notify all the closed files which are going to be reprocessed only to immediately have the rug yanked out from them anyways.

So basically we already went the other direction, and I just didn't update OnProjectReloaded.

**Comment by jasonmalinowski** [comment]
What's silly too is to close a document requires making new text loaders and such, so it's actually work to call the code to pass data that we'll immediately ignore. :smile:

**Comment by CyrusNajmabadi** [comment]
> This just sounds like a bug per @KirillOsenkov's description. Prior to 16.0 it was a requirement from the workspace that we'd have to close all the documents from a project before removing it; I removed that entirely since:

Alrighty!

**Comment by KirillOsenkov** [comment]
Whatever we end up with we should also have clear guidance for hosts about what to do in this situation.

We could manually loop over the open documents that are part of the project being reloaded, and for those that disappear from the new ProjectInfo that we're about to pass to the workspace perhaps we should close the actual files in the host, which would propagate to the workspace as document closes. Or we could keep the files, and just call close on the workspace ourselves such that the Miscellaneous Files Workspace picks them up instead.

I guess the inverse of this situation is also interesting. Suppose we have an open file in the IDE that is currently not part of any project and it got added to the Misc Workspace. Now the project is being reloaded and the new snapshot adds that document to the project. We should be claiming that open file and switch it from the Miscellaneous Files Workspace to the real one.

I wasn't sure what's the ideal interplay of the workspace and host is here, so this issue is also a request to clarify the design.

I guess whoever was the tester/QA on the Workspace feature back in 2011 didn't think this through. Luckily I'm here now to sort this all out!

Thanks!

**Comment by CyrusNajmabadi** [comment]
Closing out.
